### ISSUE TITLE
Add UnboundedStablePriorityMailbox

### ISSUE THREAD

**Issue by yahyanajar** [akkadotnet/akka.net#2652]
"UnboundedStablePriorityMailbox": an unbounded mailbox that allows for prioritization of its contents and preserves ordering for messages of equal priority

**Comment by yahyanajar** [comment]
I need a mailbox that allows for prioritization of its contents and preserves ordering for messages of equal priority. I am aware that, in akka .NET, I need to implement "UnboundedPriorityMailbox" to have a mailbox that allows for prioritization of its contents; however this latter would not preserve a FIFO ordering for messages of equal priority.

Am I missing something? Is there any workaround?

**Comment by Danthar** [comment]
@yahyanajar Have you read the docs regarding the UnboundedPriority Mailbox ?

http://getakka.net/docs/working-with-actors/Mailbox

If you implement a PriorityGenerator you should have sufficient control to achieve what you want.

**Comment by Aaronontheweb** [comment]
@danthar the key bit is this:

> "UnboundedStablePriorityMailbox": an unbounded mailbox that allows for prioritization of its contents and **preserves ordering for messages of equal priority**

Currently not implemented; will need a different priority queue implementation than what we have at the moment.

**Comment by Danthar** [comment]
> Messages ordered by the same number will remain in delivery order.

Thats whats in the docs. Seems to me thats exactly whats requested ?

**Comment by maurelio1234** [comment]
Yes it is in the docs but that's not how it works, since it is implemented
as a simple priority heap that is not stable. Would you like us to provide
a simple code that shows the t doesn't work ?


Le ven. 19 mai 2017 à 12:31, Arjen Smits <notifications@github.com> a
écrit :

**Comment by maurelio1234** [comment]
Hi,

there goes a sample code illustrating the obtained/expected message handling order, and also showing that it comes from the implementation of `ListPriorityQueue`.

https://gist.github.com/maurelio1234/724df2a39058b12730fc63d49b681a32

**Comment by spankr** [comment]
The current documentation/comment on [UnboundedPriorityMailbox](https://github.com/akkadotnet/akka.net/blob/v1.3/src/core/Akka/Dispatch/Mailbox.cs) should probably be updated.
The current comment states,
_"Messages ordered by the same number will remain in delivery order."_

akka.io's documentation states
_"Delivery order for messages of equal priority is undefined"_

**Comment by spankr** [comment]
I'm looking at [ListPriorityQueue](https://github.com/akkadotnet/akka.net/blob/v1.3/src/core/Akka/Util/ListPriorityQueue.cs) and one thing I'm not following is if it is sorting on insert (Enqueue), why is it sorting on Dequeue?

I'm also not following the logic (in Dequeue) of 

1. Take first item off the list
1. Copy last item in the list to the first position *(this violates delivery order on equal items)*
1. Truncate last item off the list.
1. Sort items

Why can't we just do: (assuming there is something to dequeue)

```
var frontItem = _data[0];   // fetch the front
_data.RemoveAt(0);
return frontItem;
```

**Comment by maurelio1234** [comment]
Maybe I'm wrong, but it's looks like a [heap implementation](http://cs.nyu.edu/courses/spring12/CSCI-GA.3033-014/Assignment3/heap.html), so your step 4 is a little bit more complicated than "sorting" elements.

The 'solution' would be either using the order of arrival of messages along with their priority to compare element with the same priority, or something like the [StablePriorityQueue ](https://github.com/akka/akka/blob/9d2bec7f232b628cc087231af75e457072823e61/akka-actor/src/main/scala/akka/util/StablePriorityQueue.scala)they have in scala Akka.

**Comment by spankr** [comment]
Yeah, its using a heap implementation so its not really a true sort. akka.io says that ordering is indeterminite (in their implementation) so the comment here should probably state something similar? 
It looks like we will likely make our own implementation that honors delivery order.

**Comment by Aaronontheweb** [comment]
closed via #3536
